File: problems/leetcode/33.search-in-rotated-sorted-array.py
#
# @lc app=leetcode id=33 lang=python3
#
# [33] Search in Rotated Sorted Array
#


# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        nums_len = len(nums)
        l = 0
        r = nums_len - 1
        last = nums[r]
        mid = (l + r) // 2
        min__num = -1
        while l < r:
            if nums[mid] > nums[(mid + 1) % nums_len]:
                min__num = (mid + 1) % nums_len
            if last > nums[mid]:
                r = mid - 1
            else:
                l = mid + 1
            last = nums[mid]
            mid = (l + r) // 2

        min__num = mid if min__num == -1 else min__num

        if nums[min__num] < target:
            return self.binary_s(min__num, nums_len - 1, target, nums)
        else:
            return self.binary_s(0, min__num - 1, target, nums)

    def binary_s(self, l, r, target, nums):
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] > target:
                r = mid - 1
            else:
                l = mid + 1
        return -1


# 1 2 4 5
# 1 2
# @lc code=end

File: problems/leetcode/33.search-in-rotated-sorted-array.go
package main

/*
 * @lc app=leetcode id=33 lang=golang
 *
 * [33] Search in Rotated Sorted Array
 */

// @lc code=start
func search(nums []int, target int) int {
	k, n := 0, len(nums)-1
	if nums[k] <= nums[n]{
		return bs(nums, k, n, target)
	}else {
		pivot := bs_piv(nums, k, n)
		if target <= nums[n]{
			return bs(nums, pivot + 1, n, target)
		}else {
			return bs(nums, k, pivot, target)
		}
		
	}
}

func bs(nums []int, min, max, target int) int{
	if max >= min{
		mid := (max + min) / 2
		
		if nums[mid] == target{
			return mid
		}else if target < nums[mid]{
			return bs(nums, min, mid-1, target)
		}else{
			return bs(nums, mid+1, max, target)
		}
	}
	return -1
}

func bs_piv(nums []int, min, max int) int{
	n := len(nums)-1
	mid := (max + min) / 2
	if mid+1 == len(nums) || nums[mid] > nums[mid+1]{
		return mid
	}else if nums[mid] < nums[n]{
		return bs_piv(nums, min, mid-1)
	}else{
		return bs_piv(nums, mid+1, max)
	}
}



// @lc code=end

